File: src/decoders/user-authorizations.ts
import type { TabLocation, UserAuthorizations } from "~/models";

export const decodeUserAuthorizations = (data: any, tabs: any[]): UserAuthorizations => {
  const canReadDiscussions = data.AvecDiscussion ?? false;
  const canDiscuss = canReadDiscussions && !(data.discussionInterdit ?? false);
  const canDiscussWithStaff = canDiscuss && (data.AvecDiscussionPersonnels ?? false);
  const canDiscussWithParents = canDiscuss && (data.AvecDiscussionParents ?? false);
  const canDiscussWithStudents = canDiscuss && (data.AvecDiscussionEleves ?? false);
  const canDiscussWithTeachers = canDiscuss && (data.AvecDiscussionProfesseurs ?? false);

  const locations: TabLocation[] = [];

  if (tabs.length > 0) {
    const traverse = (obj: any): void => {
      if ("G" in obj) {
        locations.push(obj.G);
      }

      if ("Onglet" in obj) {
        obj.Onglet.forEach(traverse);
      }
    };

    tabs.forEach(traverse);
  }

  return {
    canReadDiscussions,
    canDiscuss,
    canDiscussWithStaff,
    canDiscussWithParents,
    canDiscussWithStudents,
    canDiscussWithTeachers,

    hasAdvancedDiscussionEditor: data.AvecDiscussionAvancee ?? false,
    maxAssignmentFileUploadSize: data.tailleMaxRenduTafEleve,

    tabs: locations
  };
};

File: library/src/commonMain/kotlin/decoders/user-authorizations.kt
package decoders

import kotlinx.serialization.json.*
import models.TabLocation
import models.UserAuthorizations

fun decodeUserAuthorizations (data: JsonObject, tabs: JsonArray): UserAuthorizations {
    val canReadDiscussions = data["AvecDiscussion"]?.jsonPrimitive?.boolean ?: false
    val canDiscuss = canReadDiscussions && !(data["discussionInterdit"]?.jsonPrimitive?.boolean ?: false)
    val canDiscussWithStaff = canDiscuss && !(data["AvecDiscussionPersonnels"]?.jsonPrimitive?.boolean ?: false)
    val canDiscussWithParents = canDiscuss && !(data["AvecDiscussionParents"]?.jsonPrimitive?.boolean ?: false)
    val canDiscussWithStudents = canDiscuss && !(data["AvecDiscussionEleves"]?.jsonPrimitive?.boolean ?: false)
    val canDiscussWithTeachers = canDiscuss && !(data["AvecDiscussionProfesseurs"]?.jsonPrimitive?.boolean ?: false)

    val locations: MutableList<TabLocation> = mutableListOf()

    if (tabs.isNotEmpty()) {
        fun traverse (obj: JsonObject): Unit = run {
            if (obj.containsKey("G")) {
                val location = TabLocation.fromInt(obj["G"]!!.jsonPrimitive.int)

                if (location != null)
                    locations.add(location)
            }

            if (obj.containsKey("Onglet"))
                obj["Onglet"]!!.jsonArray.forEach{ traverse(it.jsonObject) }
        }

        tabs.forEach { traverse(it.jsonObject) }
    }

    return UserAuthorizations(
        canReadDiscussions = canReadDiscussions,
        canDiscuss = canDiscuss,
        canDiscussWithStaff = canDiscussWithStaff,
        canDiscussWithParents = canDiscussWithParents,
        canDiscussWithStudents = canDiscussWithStudents,
        canDiscussWithTeachers = canDiscussWithTeachers,

        hasAdvancedDiscussionEditor = data["AvecDiscussionAvancee"]?.jsonPrimitive?.boolean ?: false,
        maxAssignmentFileUploadSize = data["tailleMaxRenduTafEleve"]!!.jsonPrimitive.long,

        tabs = locations
    )
}